### ISSUE TITLE
Connection to Emby server fails partially

### ISSUE THREAD

**Issue by mammo0** [MediaBrowser/plugin.video.emby#308]
Hello,

since last week the the Emby Kodi addon couldn't fully connect to my Emby server instance. It appears that the first ping/connection is successful and the addon can logon at the server, but then a Python error happens (in the simplejson module).
There are no errors on server side.

After that no interaction between the addon and the Emby server is possible (no library update, movie/tv-show playing, etc.).

I did neither update the Kodi addon nor the Emby server version. The log shows version 18.3 of Kodi. But the error also occurs on 18.4. On my other Kodi device (OS X) there is no such error... So I think it's not a Problem with my server settings.

```
2019-09-29 15:03:50.023 T:140100541339392  NOTICE: EMBY.hooks.webservice -> --->[ webservice/57578 ]
2019-09-29 15:03:50.024 T:140101623473920  NOTICE: EMBY.patch -> --->[ patch ]
2019-09-29 15:03:50.024 T:140101623473920  NOTICE: EMBY.patch -> --[ check updates/181167259 ]
2019-09-29 15:03:50.322 T:140101623473920  NOTICE: EMBY.patch -> --<[ objects/181167259 ]
2019-09-29 15:03:50.393 T:140101623473920  NOTICE: EMBY.service -> -->[ service ]
2019-09-29 15:03:50.393 T:140101623473920  NOTICE: EMBY.service -> Delay startup by 0 seconds.
2019-09-29 15:03:50.693 T:140100532418304  NOTICE: EMBY.entrypoint.service -> --->>>[ EMBY ]
2019-09-29 15:03:50.693 T:140100532418304  NOTICE: EMBY.entrypoint.service -> Version: 4.1.14
2019-09-29 15:03:50.696 T:140100532418304  NOTICE: EMBY.entrypoint.service -> KODI Version: 18.3 Git:20190624-nogitfound
2019-09-29 15:03:50.720 T:140100532418304  NOTICE: EMBY.entrypoint.service -> Platform: Linux
2019-09-29 15:03:50.722 T:140100532418304  NOTICE: EMBY.entrypoint.service -> OS: Kodi/7795MB
2019-09-29 15:03:50.722 T:140100532418304  NOTICE: EMBY.entrypoint.service -> Python Version: 2.7.16 (default, Mar 11 2019, 18:59:25) 
                                            [GCC 8.2.1 20181127]
2019-09-29 15:03:50.723 T:140100532418304  NOTICE: EMBY.entrypoint.service -> Using dynamic paths: True
2019-09-29 15:03:50.724 T:140100532418304  NOTICE: EMBY.entrypoint.service -> Log Level: 1
2019-09-29 15:03:50.725 T:140100532418304  NOTICE: EMBY.entrypoint.service -> ---[ db/None ]
2019-09-29 15:03:50.751 T:140100532418304  NOTICE: Emby -> ---[ START EMBYCLIENT ]---
2019-09-29 15:03:50.772 T:140100532418304  NOTICE: EMBY.views -> Unable to retrieve libraries: NotConnected
2019-09-29 15:03:50.783 T:140100532418304  NOTICE: EMBY.client -> DeviceId loaded: DA06F2417A154BCABABE762FF15C336D
2019-09-29 15:03:50.784 T:140102235846400  NOTICE: EMBY.hooks.monitor -> --->[ listener ]
2019-09-29 15:03:50.784 T:140101749298944  NOTICE: Emby -> ---[ STOPPED EMBYCLIENT: default ]---
2019-09-29 15:03:50.784 T:140101749298944  NOTICE: EMBY.connect -> --[ server/default ]
2019-09-29 15:03:50.785 T:140101749298944  NOTICE: Emby -> ---[ START EMBYCLIENT ]---
2019-09-29 15:03:50.786 T:140101749298944  NOTICE: Emby.emby.core.connection_manager -> Begin connect
2019-09-29 15:03:51.788 T:140101749298944  NOTICE: Emby.emby.core.connection_manager -> Found Servers: []
2019-09-29 15:03:51.789 T:140101749298944  NOTICE: Emby.emby.core.connection_manager -> Begin connectToServers, with 1 servers
2019-09-29 15:03:51.789 T:140101749298944  NOTICE: Emby.emby.core.connection_manager -> testing connection mode 2 with server emby
2019-09-29 15:03:51.789 T:140101749298944  NOTICE: Emby.emby.core.connection_manager -> tryConnect url: https://{emby-server}/emby/system/info/public
2019-09-29 15:03:51.951 T:140101749298944  NOTICE: Emby.emby.core.connection_manager -> calling onSuccessfulConnection with connection mode 2 with server emby
2019-09-29 15:03:52.096 T:140101749298944  NOTICE: Emby.emby.client -> User is authenticated.
2019-09-29 15:03:52.420 T:140101749298944  NOTICE: EMBY.setup -> ---<[ setup ]
2019-09-29 15:03:52.422 T:140101732513536  NOTICE: Emby.emby.core.ws_client -> Websocket url: wss://{emby-server}/embywebsocket?api_key={emby-token}&device_id=DA06F2417A154BCABABE762FF15C336D
2019-09-29 15:03:52.423 T:140100558124800  NOTICE: EMBY.library -> --->[ library ]
2019-09-29 15:03:52.501 T:140101749298944  NOTICE: EMBY.hooks.monitor -> -->[ q:monitor/LoadServer ]
2019-09-29 15:03:52.501 T:140101749298944  NOTICE: EMBY.hooks.monitor -> --[ post capabilities/1592601c27f5460b99ee9778982da08a ]
2019-09-29 15:03:52.610 T:140101732513536  NOTICE: Emby.emby.core.ws_client -> --->[ websocket ]
2019-09-29 15:03:52.667 T:140101749298944  NOTICE: EMBY.hooks.monitor -> ERROR:: 'NoneType' object is not callable
                                            Traceback (most recent call last):
                                              File "/<user_home>/.kodi/addons/plugin.video.emby/resources/lib/hooks/monitor.py", line 503, in run
                                                func(server, data)
                                              File "/<user_home>/.kodi/addons/plugin.video.emby/resources/lib/hooks/monitor.py", line 324, in LoadServer
                                                self.server_instance(data['ServerId'])
                                              File "/<user_home>/.kodi/addons/plugin.video.emby/resources/lib/hooks/monitor.py", line 158, in server_instance
                                                self.post_capabilities(server)
                                              File "/<user_home>/.kodi/addons/plugin.video.emby/resources/lib/hooks/monitor.py", line 195, in post_capabilities
                                                'IconUrl': "https://raw.githubusercontent.com/MediaBrowser/plugin.video.emby/master/kodi_icon.png",
                                              File "/<user_home>/.kodi/addons/plugin.video.emby/libraries/emby/core/api.py", line 315, in post_capabilities
                                                return  self.sessions("/Capabilities/Full", "POST", json=data)
                                              File "/<user_home>/.kodi/addons/plugin.video.emby/libraries/emby/core/api.py", line 73, in sessions
                                                return  self._post("Sessions%s" % handler, json, params)
                                              File "/<user_home>/.kodi/addons/plugin.video.emby/libraries/emby/core/api.py", line 56, in _post
                                                return  self._http("POST", handler, {'params': params, 'json': json})
                                              File "/<user_home>/.kodi/addons/plugin.video.emby/libraries/emby/core/api.py", line 50, in _http
                                                return  self.client.request(request)
                                              File "/<user_home>/.kodi/addons/plugin.video.emby/libraries/emby/core/http.py", line 171, in request
                                                response = r.json()
                                              File "/<user_home>/.kodi/addons/plugin.video.emby/libraries/requests/models.py", line 808, in json
                                                return complexjson.loads(self.text, **kwargs)
                                              File "/usr/lib/python2.7/site-packages/simplejson/__init__.py", line 538, in loads
                                                return _default_decoder.decode(s)
                                              File "/usr/lib/python2.7/site-packages/simplejson/decoder.py", line 370, in decode
                                                obj, end = self.raw_decode(s)
                                              File "/usr/lib/python2.7/site-packages/simplejson/decoder.py", line 400, in raw_decode
                                                return self.scan_once(s, idx=_w(s, idx).end())
                                              File "/usr/lib/python2.7/site-packages/simplejson/errors.py", line 41, in __init__
                                                ValueError.__init__(self, errmsg(msg, doc, pos, end=end))
                                            TypeError: 'NoneType' object is not callable
2019-09-29 15:03:53.516 T:140100558124800  NOTICE: EMBY.database -> [emby] 1 rows updated.
2019-09-29 15:03:53.904 T:140100558124800  NOTICE: Previous line repeats 2 times.
2019-09-29 15:03:53.904 T:140100558124800  NOTICE: EMBY.helper.utils -> exiiiiitttinggg
2019-09-29 15:03:53.904 T:140100558124800  NOTICE: EMBY.library -> ---<[ library ]
```

**Comment by mammo0** [comment]
Today I had some time to analyze the error.

After a lot of trial and error I think there must be a bug in my Python-Kodi environment. After the method call in line 400 of the file `/usr/lib/python2.7/site-packages/simplejson/decoder.py`
```
...
return self.scan_once(s, idx=_w(s, idx).end())
```
the Python interpreter jumps to the file `/usr/lib/python2.7/site-packages/simplejson/errors.py`. I don't know what black magic is happening here, but when the interpreter is in this file all my global variables from globals() are `None`. Before the jump to that file everythiung is OK and my global variables are all there. I found nothing on the internet that describes this problem.

So because of all global Variables are `None` the `ValueError` class from `__builtin__` is also `None`. So the call in the file `/usr/lib/python2.7/site-packages/simplejson/errors.py`
```
...
ValueError.__init__(self, errmsg(msg, doc, pos, end=end))
...
```
fails of course with the message `TypeError: 'NoneType' object is not callable`.

I've got no clue why this is happening...

But I found a workaround for my problem. Because the main problem is that in the file `plugin.video.emby/libraries/emby/core/http.py` the method `json()` is called on a `204-Request` object. So the content of the server response should be parsed, even if there is no content, because of the `204`-HTTP-Response. This results in the call of `complexjson.loads(self.text, **kwargs)` in file `plugin.video.emby/libraries/requests/models.py`, where `self.text` is an empty string.
Trying to run `simplejson.loads('')` in a normal Python2.7 console environment gives me also an error:
```
simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```
So I think that there is no need to call the `json()` method on the response object in Emby if it is empty anyway.

I wrote a little patch that the `json()` method should not be called on an empty response object:
```
diff --git a/libraries/emby/core/http.py b/libraries/emby/core/http.py
index 00fbd382..e1786761 100644
--- a/libraries/emby/core/http.py
+++ b/libraries/emby/core/http.py
@@ -168,6 +168,9 @@ class HTTP(object):
 
                 try:
                     self.config['server-time'] = r.headers['Date']
+                    if r.status_code == 204:
+                        # return, because there is no response
+                        return
                     response = r.json()
                     LOG.debug(json.dumps(response, indent=4))
```
It basically checks for a `204`-HTTP-Response code and simple returns from the method, like if a `ValueError` occurs during JSON parsing (see the try-catch block).
